src/test/bin/winkernel/control.cpp

```diff
@@ -585,6 +587,76 @@ typedef union {
     X; \
     Status = Client->TestFailure ? STATUS_FAIL_FAST_EXCEPTION : STATUS_SUCCESS;
 
+// Base template providing a readable error for unsupported scenarios
+template<class... Args>
+QUIC_STATUS InvokeTestFunction(void(Args...), const uint8_t*, uint32_t) {
+    static_assert(false, "Only functions with no argument or one constant reference argument are supported");
+}
+
+// Specialization for functions with one const reference argument
+template<class Arg>
+QUIC_STATUS InvokeTestFunction(void(*func)(const Arg&), const uint8_t* argBuffer, uint32_t argBufferSize) {
+    if (sizeof(Arg) != argBufferSize) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "Invalid parameter size for test function");
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    const Arg& arg = *reinterpret_cast<const Arg*>(argBuffer);
+    func(arg);
+    return QUIC_STATUS_SUCCESS;
+}
+
+// Specialization for functions with no arguments
+template<>
+QUIC_STATUS InvokeTestFunction(void(*func)(), const uint8_t*, uint32_t argBufferSize) {
+    if (0 != argBufferSize) {
+        QuicTraceEvent(
+            LibraryError,
+            "[ lib] ERROR, %s.",
+            "Parameter provided for a test function expecting none");
+        return QUIC_STATUS_INVALID_PARAMETER;
+    }
+
+    func();
+    return QUIC_STATUS_SUCCESS;
+}
+
+#define RegisterTestFunction(Function) \
+    do { \
+        if (strcmp(Request->FunctionName, #Function) == 0) { \
+            return InvokeTestFunction( \
+                Function, \
+                (const uint8_t*)(Request + 1), \
+                Request->ParameterSize); \
+        } \
+    } while (false)
+
+QUIC_STATUS
+ExecuteTestRequest(
+    _In_ QUIC_RUN_TEST_REQUEST* Request
+    )
+{
+    // Ensure null termination
+    Request->FunctionName[sizeof(Request->FunctionName) - 1] = '\0';
+
+    // Register any test functions here
```

Comment:
aww, so we still have to register the test function with the driver? :(
At least it's better than having to get the IOCTL code correct AND update the switch statement list.